rango-exchange/rango-client
# feat: add logging

## How it works

This is an experimental version of solution for our logging

Here is the high level architecture:

<img width="1203" alt="arch" src="https://github.com/rango-exchange/rango-client/assets/120930547/9cd43097-e514-4e7b-b884-4ffb26f30578">


*core*: expose `error`, `warn`, ... can be used by any library or client to emit log records.
*subscriber*: All the log records will be captured by subscriber and will be passed to `layers` to process them.
*console* and *sentry*: Two separate layers to process the logs and one, print to console and the other one send them to sentry.

You can freely record logs everywhere you wanted. just import `core` and use the methods. if there is no subscriber to capture them, nothing happens.


## How to test

It currently happens only on rejecting transactions (logRpc function) from wallet. if you want to test more levels, take a look at this commit: https://github.com/rango-exchange/rango-client/pull/601/commits/e17844b98b3fc00d6085b2ed705d10218326e4f3

There is another PR on [our dApp repo](https://github.com/rango-exchange/app/pull/61), you can see how you can link and run some tests against this PR using dApp.

I also added a subscriber to widget, only on dev mode (localhost). It will be eliminated on production and there is no plan to add it to production yet. So you can run `yarn dev:widget` to test PR as well.



## Next steps

We can add more logging our libs, but one important thing is every packages should have their own error instances (like `RangoPresetError` in this PR), it enriches the log records.


## TODOs

- [x] Initial draft
- [x] Console layer
- [x] Max level filtering
- [ ] Set Max level from env or window
- [x] Sentry Layer
- [x] Customized instance of error for our libraries 
- [x] Caret version?
